### ISSUE TITLE
strlen of NULL string segfaults in nvmet.c

### ISSUE THREAD

**Issue by pkoutoupis** [pkoutoupis/rapiddisk#155]
When not running with json switch, NULL is fed into some of the nvme target functions for the return message string.
```
                        if (json_flag) {
                                rc = nvmet_export_volume(disk, cache, backing, host, port, generic_msg);
                                print_message(rc, generic_msg, json_flag);
                        } else {
                                rc = nvmet_export_volume(disk, cache, backing, host, port, NULL);
                        }

```
When a failure occurs due to an inappropriate invocation of rapiddisk without the nvme modules installed and it check the strlen of a NULL value:
```
        if (port != INVALID_VALUE) {
                ports = nvmet_scan_all_ports(return_message);
                if ((ports == NULL) && (strlen(return_message) != 0)) {
                        msg = "%s";
                        print_error(msg, return_message, return_message);
                        return INVALID_VALUE;
                }
```
This is NOT a fix or workaround but gives me an idea of the problem:
```
diff --git a/src/nvmet.c b/src/nvmet.c
index d16402b..2a7311c 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -533,6 +533,8 @@ int nvmet_export_volume(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char *d
        struct dirent **list;
        char *msg;
        struct NVMET_PORTS *ports;
+       char error_message[NAMELEN] = {0};
+
 
        /*
         * We do not care if the device has already been exported. We will continue to go through
@@ -564,10 +566,10 @@ int nvmet_export_volume(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char *d
        rc = INVALID_VALUE;
 
        if (port != INVALID_VALUE) {
-               ports = nvmet_scan_all_ports(return_message);
-               if ((ports == NULL) && (strlen(return_message) != 0)) {
+               ports = nvmet_scan_all_ports(error_message);
+               if ((ports == NULL) && (strlen(error_message) != 0)) {
                        msg = "%s";
-                       print_error(msg, return_message, return_message);
+                       print_error(msg, return_message, error_message);
                        return INVALID_VALUE;
                }
                while (ports != NULL) {
```

**Comment by matteotenca** [comment]
This is another approach... I added a check to the export function to avoid to add an already allowed host to the allowed lists... I don't know if it is useful.

I believe that some other functions suffer from the same bug you discovered!

```diff
diff --git a/src/nvmet.c b/src/nvmet.c
index 6d7f7a2..40e5a14 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -249,14 +249,14 @@ struct NVMET_PORTS *nvmet_scan_ports(char *return_message)
  */
 
 /**
- * > This function scans the /sys/kernel/config/nvmet/ports directory for all NVMe Target ports and returns a linked list
- * of all ports found
+ * This function scans the /sys/kernel/config/nvmet/ports directory for all the ports that are currently configured
  *
  * @param return_message This is a pointer to a string that will be populated with an error message if the function fails.
+ * @param rc return code
  *
  * @return A pointer to the first element of a linked list of struct NVMET_PORTS.
  */
-struct NVMET_PORTS *nvmet_scan_all_ports(char *return_message)
+struct NVMET_PORTS *nvmet_scan_all_ports(char *return_message, int *rc)
 {
 	int err, n = 0;
 	char file[NAMELEN * 2] = {0};
@@ -321,6 +321,7 @@ struct NVMET_PORTS *nvmet_scan_all_ports(char *return_message)
 		}
 	}
 	ports = clean_scandir(ports, err);
+	*rc = SUCCESS;
 	return ports_head;
 }
 
@@ -526,12 +527,12 @@ int nvmet_view_exports(bool json_flag, char *error_message)
  */
 int nvmet_export_volume(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char *device, char *host, int port, char *return_message)
 {
-	int rc = INVALID_VALUE, n, err;
+	int rc = INVALID_VALUE, n, err, rv = INVALID_VALUE;
 	FILE *fp;
 	mode_t mode = (S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
 	char hostname[0x40] = {0x0}, path[NAMELEN] = {0x0}, path2[NAMELEN] = {0x0};
 	struct dirent **list;
-	char *msg, error_message[NAMELEN] = {0};
+	char *msg;
 	struct NVMET_PORTS *ports;
 
 	/*
@@ -564,10 +565,8 @@ int nvmet_export_volume(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char *d
 	rc = INVALID_VALUE;
 
 	if (port != INVALID_VALUE) {
-		ports = nvmet_scan_all_ports(error_message);
-		if ((ports == NULL) && (strlen(error_message) != 0)) {
-			msg = "%s";
-			print_error(msg, return_message, error_message);
+		ports =  nvmet_scan_all_ports(return_message, &rv);
+		if (rv == INVALID_VALUE) {
 			return INVALID_VALUE;
 		}
 		while (ports != NULL) {
@@ -623,6 +622,25 @@ int nvmet_export_volume(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char *d
 	} else {
 		/* Configure the target to be seen only by the specified host(s) */
 
+		/* Check if the host is already in the allowed list */
+		sprintf(path, "%s/%s%s-%s/allowed_hosts", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+		if ((err = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
+			msg = "Error. Unable to access %s. %s: scandir: %s";
+			print_error(msg, return_message, path, __func__, strerror(errno));
+			return INVALID_VALUE;
+		}
+
+		for (n = 0; n < err; n++) {
+			if (strcmp(list[n]->d_name, host) == 0) {
+				msg = "Error. The host %s is already in the allowed list.";
+				print_error(msg, return_message, host);
+				list = clean_scandir(list, err);
+				return INVALID_VALUE;
+			}
+		}
+
+		list = clean_scandir(list, err);
+
 		/* Make sure that no other hosts can access the target */
 		sprintf(path, "%s/%s%s-%s/attr_allow_any_host", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
 		if ((fp = fopen(path, "w")) == NULL){
@@ -1053,11 +1071,11 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
 int nvmet_view_ports(bool json_flag, char *error_message)
 {
 	int i = 1;
-	int rc = SUCCESS;
+	int rc = SUCCESS, rv = INVALID_VALUE;
 	struct NVMET_PORTS *ports = NULL, *tmp_ports = NULL;
 
-	ports = nvmet_scan_all_ports(error_message);
-	if ((ports == NULL) && (strlen(error_message) != 0)) {
+	ports = nvmet_scan_all_ports(error_message, &rv);
+	if (rv == INVALID_VALUE) {
 		return INVALID_VALUE;
 	}
 
@@ -1308,11 +1326,11 @@ int nvmet_view_exports_json(char *error_message, char **json_result) {
  * @param json_result This is the JSON string that will be returned to the caller.
  */
 int nvmet_view_ports_json(char *error_message, char **json_result) {
-	int rc = SUCCESS;
+	int rc = SUCCESS, rv = INVALID_VALUE;
 	struct NVMET_PORTS *ports = NULL;
 
-	ports = nvmet_scan_all_ports(error_message);
-	if ((ports == NULL) && (strlen(error_message) != 0)) {
+	ports = nvmet_scan_all_ports(error_message, &rv);
+	if (rv == INVALID_VALUE) {
 		return INVALID_VALUE;
 	}
 
diff --git a/src/nvmet.h b/src/nvmet.h
index a4fc291..f871e8e 100644
--- a/src/nvmet.h
+++ b/src/nvmet.h
@@ -41,7 +41,7 @@ SPDX-License-Identifier: GPL-2.0-or-later
 
 struct NVMET_PROFILE *nvmet_scan_subsystem(char *return_message);
 struct NVMET_PORTS *nvmet_scan_ports(char *return_message);
-struct NVMET_PORTS *nvmet_scan_all_ports(char *return_message);
+struct NVMET_PORTS *nvmet_scan_all_ports(char *return_message, int *rc);
 char *nvmet_interface_ip_get(char *interface, char *return_message);
 #ifndef SERVER
 int nvmet_view_exports(bool json_flag, char *error_message);
```

**Comment by pkoutoupis** [comment]
@matteotenca Regarding the `nvmet_scan_all_ports()` change, I don't know that I agree with this approach. Let me explain below....

By adding an `rc` to check that the scanning of ports went through, you miss the opportunity to print a message in stdout if the return_message string is NULL. For example:
```
-		ports = nvmet_scan_all_ports(error_message);
-		if ((ports == NULL) && (strlen(error_message) != 0)) {
-			msg = "%s";
-			print_error(msg, return_message, error_message);
+		ports =  nvmet_scan_all_ports(return_message, &rv);
+		if (rv == INVALID_VALUE) {
 			return INVALID_VALUE;
 		}
```
msg will still be set to `error_message` which should be a valid string array of proper size. When passed into `print_error()`, you check if `return_message` is NULL and if so, you print the format + args to stdout otherwise, it is copied to `return_message`. I rather like that original approach you conceived. By checking a return code (rv / rc) passed into it, you avoid that and miss the opportunity to tell users that the NVMe Target drivers aren't loaded or something else. Am I misinterpreting something?

Regarding host check, I need to think this through a few scenarios and I will get back to you.

**Comment by matteotenca** [comment]
> msg will still be set to `error_message` which should be a valid string array of proper size. When passed into `print_error()`, you check if `return_message` is NULL and if so, you print the format + args to stdout otherwise, it is copied to `return_message`. I rather like that original approach you conceived. By checking a return code (rv / rc) passed into it, you avoid that and miss the opportunity to tell users that the NVMe Target drivers aren't loaded or something else. Am I misinterpreting something?

I took into consideration the very same topic you noticed, and that's why I discarded a first implementation I made. But in this one, since the `error_message` pointer is passed to the `nvmet_scan_all_ports()` function as it is received by `nvmet_export_volume()`, the buffer it refers to (if it is not `NULL`) is filled by the `print_error()` calls INSIDE `nvmet_scan_all_ports()`. If `error_message` is `NULL`, the message is printed to console correctly. The `rc/rv` return code is used inside `nvmet_export_volume()` just to `return` upon error, without taking into consideration the `error_message` value, since I am sure this job was aleady done inside `nvmet_scan_all_ports()`.

Long story short: the previous behaviour is basically kept, but the `error_message` buffer is valorized, if not `NULL`, inside the  `nvmet_scan_all_ports()` function. The previous implementation was reduntant: line 570 is useless, since if `error_message` is not `NULL`, the calls rewrites the content of the buffer to itself. 

https://github.com/pkoutoupis/rapiddisk/blob/f45fd05526b43811c105875492ae775d087aaa68/src/nvmet.c#L568-L572

Furthermore, passing `error_message` to `print_error()` two times, causes another issue, as stated into `printf` man page. At line 106, the `vsprintf` call becomes something like `vsprintf(error_message, msg, error_message)`.

`printf()` documentation:

```
Some programs imprudently rely on code such as the following

           sprintf(buf, "%s some further text", buf);

       to append text to buf.  However, the standards explicitly note
       that the results are undefined if source and destination buffers
       overlap when calling sprintf(), snprintf(), vsprintf(), and
       vsnprintf().  Depending on the version of gcc(1) used, and the
       compiler options employed, calls such as the above will not
       produce the expected results.
```

https://github.com/pkoutoupis/rapiddisk/blob/f45fd05526b43811c105875492ae775d087aaa68/src/rdsk.c#L102-L112

So, what is really missing in my patch are two calls to `print_error()` after lines 297 and 309:

https://github.com/pkoutoupis/rapiddisk/blob/f45fd05526b43811c105875492ae775d087aaa68/src/nvmet.c#L291-L310

Since I wanted to keep `NULL` as a valid return value from `nvmet_scan_all_ports()`, the same as you desidered if I got the point, I decided to passing `rv` as a reference to be able to gather the exit status AND the return value. Another option would have been to receive two values, but I believe this cannot be done in C in a straightforward way (as in Python), or to create a new struct which could contain both the exit status and the return value, but this seemed really awkyard to me.

Another choice could have been to return `INVALID_VALUE` from  `nvmet_scan_all_ports()` upon error, but this would have required a cast from `NVMET_PORTS *` to `int` to perform the check.

So I believe that, after adding those missing calls in `nvmet_scan_all_ports()`, this approach is able to preserve the `print_error()` functionality, to log all the messages properly AND to preserve `NULL` as a valid return value.

I hope I got the point and explained well!
 
> Regarding host check, I need to think this through a few scenarios and I will get back to you.
> 
> **EDIT:** I already check:
> More specifically, this if statement checks if the host is already listed in the directory:
> 
> ```
> if (access(path2, F_OK) != SUCCESS) {
> ```

My concern is with filesystem operations being performed before this check anyway. And to say it all, some are performed anyway if the check I added is not moved before line 587.

But my knowledge of NVME exporting processes and kernel filesystem things is very very very poor, if not null :-) so please consider my concerns void if it is the case!

> **EDIT 2:** Actually, I don't and probably shouldn't return in the event I set the same host NQN to another port. If it is already exported through port 1 but now I want to export it through port 3.
> 
> This whole NVMe subsystem management will definitely get some restructuring. It will need to be part of another PR (#148).

I believe you are right. In my mind, these functions should do some kind of garbage collection and undo operations which became useless upon error, or check if a set of conditions are met before starting to manipulate the filesystem. But this are really my 2 cents...
 
> **EDIT 3:** I am sorry for all of the updates. I walk away and then something else pops into my head. I guess I could short circuit things and use a `goto` statement if the path already exists and then have it skip to the port stuff. Or at least to the enabling part above the port stuff (and make sure that no outside force disabled the volume at any point in time).

If you have patience, I can try to rework this function, and if you approve the result, we can change the others accordingly.

A final note: I am quite sure this issue with `print_error()` and the pointer to char which may or may not be `NULL` that cause seg faults is present in another part of the code.

All of this was my mistake!

**Comment by pkoutoupis** [comment]
OK. So, I now see and agree with the comments and rework pertaining to `nvmet_scan_all_ports()`. That aside, I would prefer to save the allowed host management for a separate PR as it will be quite a bit of rework (#148). For now, I have added a `goto` statement to avoid certain redundant tasks from reoccurring. A new commit was pushed containing this `nvmet_scan_all_ports()` rework moments ago to https://github.com/pkoutoupis/rapiddisk/pull/153.

Anyway, with your thumbs up, I can merge #153 unless something else glaring catches your eye.

**Comment by pkoutoupis** [comment]
Fixed in https://github.com/pkoutoupis/rapiddisk/pull/153.
